examples/sysl.wbnf

```diff
@@ -10,60 +10,131 @@ import  -> "import"  prefix="/"{0,2} import=PATH ( "as"  APP_NAME)? ("~" mode=NA
+    \n+ level=INDENT
+        (COMMENT | (name=NAME "<:") type=(collection_type|type_spec) optional="?"? attribs?):INDENT_SEP EOL;
 wrap -> "!wrap" NAME attribs?  ":" COMMENT*
-    INDENT
-        (table | "!table"APP_NAME) :INDENT_SEP EOL;
+    \n+ level=INDENT
+        (table | "!table"APP_NAME):INDENT_SEP EOL;
 
 
-type_spec -> (NativeDataType=("int32" | "int64" | "int" |
-                             "float" | "string" | "date" |
-                             "bool" | "decimal" | "datetime") | UserType=APP_NAME)
+type_spec -> (NativeDataTypes | UserType=APP_NAME)
              SizeSpec=("(" \d+ (array=".."|sized=".")? \d* ")")?;
 collection_type -> ("set" | "sequence") "of" type_spec;
 
 
 // -------------- Endpoint     --------------- //
 
-endpoint -> "/" [^\n:]* ":" COMMENT* INDENT [^\n]+:INDENT_SEP;
+endpoint -> (rest_endpoint | simple_endpoint) EOL;
 
+simple_endpoint -> SHORTCUT | (endpoint_name=(PATH) QSTRING? params? attribs? ":"
+                        (SHORTCUT | \n+ level=INDENT stmt:INDENT_SEP));
+
+rest_endpoint -> http_path attribs? ":" \n+ level=INDENT (method_def | rest_endpoint):INDENT_SEP;
+
+method_def -> method=(HTTP_VERBS) params? query_param? attribs? ":" \n+ level=INDENT stmt:INDENT_SEP;
+
+params -> "(" ( reference | field ):"," ")";
+
+query_param -> "?" (Name=(NAME) "=" (NativeDataTypes| NAME | "{" NAME "}") optional="?"?):"&";
+
+field -> NAME (array_size? "<:" (field_type | (\n+ level=INDENT @:INDENT_SEP)) QSTRING?)?;
+
+field_type -> collection_type | (type_spec optional="?"? attribs?) ;
 
 // --------------    views    --------------- //
 view -> "!view" NAME "(" view_params? ")"
         ("->" retType=(collection_type|type_spec))?
-        attribs? ":" ((INDENT expr_block:INDENT_SEP) | abstract="[~abstract]");
+        attribs? ":" ((\n+ level=INDENT expr_block:INDENT_SEP) | abstract="[~abstract]");
 
 view_params -> (NAME "<:" (collection_type|type_spec)):",";
 
 expr_block -> "";
 
+// --------------    stmt    --------------- //
+stmt -> (if_else
+            | for_stmt
+            | ret_stmt
+            | call_stmt
+            | one_of_stmt
+            | http_method_comment=(COMMENT)
+            | group_stmt
+            | text_stmt
+            | annotation
+        ) attribs?;
+
+if_else -> if_stmt else_stmt*;
+
+if_stmt -> "if" PREDICATE ":" \n+ level=INDENT stmt:INDENT_SEP;
+
+else_stmt -> "else" PREDICATE? ":" \n+ level=INDENT stmt:INDENT_SEP;
+
+for_stmt -> for_type=("alt" | "until" | "for each" | "for" | "loop" | "while") PREDICATE ":" \n+ level=INDENT stmt:INDENT_SEP;
+
+one_of_stmt -> "one" "of" ":" \n+ level=INDENT one_of_cases:INDENT_SEP;
+
+one_of_cases -> one_of_case_label=(QSTRING ":"| NAME ":" | TEXT_LINE ":" | ":")  \n+ level=INDENT stmt:INDENT_SEP;
+
+call_stmt -> target=("." | NAME:"::") "<-" target_endpoint=(TEXT_LINE) call_args?;
+
+call_args -> "(" ((QSTRING | NAME)+ | (NAME "<:" (NAME | NativeDataTypes))):"," ")";
```

Comment:
`(NAME | NativeDataTypes)` should be `type_spec` check elsewhere also

Reply:
`type_spec` allows `APP_NAME`, is that okay for `call_args`?

Reply:
should I change it to `type_spec -> (NativeDataTypes | UserType=APPLICATION_NAME | NAME)
             SizeSpec=("(" \d+ (array=".."|sized=".")? \d* ")")?;`